### ISSUE TITLE
Can the embedded encryption key be relocated for use with package management (OpenUPM)

### ISSUE THREAD

**Issue by bitindicium** [sabresaurus/PlayerPrefsEditor#8]
I was interested in pulling in this library via OpenUPM. There's an embed encryption key that we're recommended to change, but doing so would disallow the use of package management (I think?) Could this be exposed for runtime injection? There may be other methods that are preferable in the package manager system, such as having installation create an asset, but I'm unfamiliar.

**Comment by rfadeev** [comment]
Hi,

This is a know limitation of the package, unfortunately (mentioned in #3). The easiest way to change the encryption key would be having local copy of the package. Yes, no OpenUPM for it then.

I have been thinking about the ways of managing encryption but have not found a good solution. I prototyped the option to do so via [Unity's settings manager](https://docs.unity3d.com/Packages/com.unity.settings-manager@1.0/manual/index.html) and put encryption key into auto-generated C# file in Assets folder, but bumped into the problem that package itself cannot use main project source code - since it is a separate assembly.

This leaves the option of having some asset with encryption key which can be changed. Then the problem of loading that asset at runtime and tracking key changes in it at editor time occurs. Right now with key living in the code no such problem exists.

Pretty much the problem is bound to the lack of editor and runtime friendly settings API for packages.

@sabresaurus Could you please provide your input? Maybe I am missing some obvious solution here.

**Comment by bitindicium** [comment]
If there isn't a better solution supported by the package manager system, could a workaround like the following work?

```
    public static class SimpleEncryption
    {
        // IMPORTANT: Make sure to change this key for each project you use this encryption in to help secure your
        // encrypted values. This key must be exactly 32 characters long (256 bit).
        private static string key = ":{j%6j?E:t#}G10mM%9hp5S=%}2,Y26C";

        public static string Key
        {
           set
           {
              key = value;
              SetupProvider();
           }
        }
```

And in user code:

```
using Sabresaurus.PlayerPrefsUtilities;
using UnityEngine;

public static class Bootstrap
{
   [RuntimeInitializeOnLoadMethod(RuntimeInitializeLoadType.BeforeSceneLoad)]
   private static void Bootstrap_BeforeInitialSceneLoad()
   {
      SimpleEncryption.Key = ":{j%6j?E:t#}G10mM%9hp5S=%}2,Y26C";
   }
}
```

**Comment by rfadeev** [comment]
Thanks for sharing the idea! Yes, this could be a workaround but it has a limitation of Player Prefs editor window (Window -> PlayerPrefs Editor) functionality of encrypting/decrypting not working properly for a custom key. Fixing that would require some `BootstrapEditor` class with `InitializeOnLoad`  to set key for editor time.

Maybe package should provide an editor utility to generate code for`Bootstrap`/`BootstrapEditor` and new key value.

**Comment by sabresaurus** [comment]
> bumped into the problem that package itself cannot use main project source code - since it is a separate assembly.

Technically you can reference Assets asmdefs from Packages, it would just be horrible to do so!

I agree that using a combination of RuntimeInitializeOnLoad and InitalizeOnLoad is a good workaround for this issue, I would like to find a way to do it that doesn't rely on a boilerplate class being created, but I'm struggling to think of a way of doing it without bringing in lots of dependencies that seem like overkill or messing with build pipelines. I feel like there should be a simpler way of doing this but it's currently eluding me. I'll do some more digging and see if there's another API that may help

**Comment by sabresaurus** [comment]
Dug into this some more and I couldn't see a good cross-platform way of embedding keys into builds without them being trivial to extract (e.g. from resources/streaming assets). Hardcoding them (where they'll end up in IL2CPP for most build paths) seems like a reasonably secure way of stopping trivial extraction, though obviously anything can be reverse engineered when the player has the build.

I've adapted your suggestions @bitindicium @rfadeev which is now in main - there is now a button on the editor to create a script file with a random key:

![image](https://user-images.githubusercontent.com/17784523/146953479-1bd1d1e4-c439-41b3-b2f8-8eb0e4f9e9f8.png)

**Comment by bitindicium** [comment]
Thank you. Minor follow up. The first time I ran this it produced:
```
public override string GetCustomKey()
{
    return "v#2N<6bJ8?6>wt/l~@B}GVt/kbEI""jS"; 
}
```

Which failed to compile due to the embedded double quotes. Maybe these need to be escaped or disallowed.

I re-ran and got another key which I'm using now though. Thanks again!

**Comment by rfadeev** [comment]
@bitindicium Hey, good catch with double quotes problem!

I would say disallow them for simplicity. Even though adding escaping can be done like:
```csharp
                        var keyCharacter = availableCharacters[index];
                        if (keyCharacter == '\"')
                        {
                            key += "\\\"";
                        }
                        else
                        {
                            key += keyCharacter;
                        }
```

@sabresaurus What do you think?

**Comment by sabresaurus** [comment]
Thanks, I have fixed this in 27b6ad4b1139d611a7dfb0e31d215f156d587a2b by changing the template code to work with byte arrays, rather than using a subset of characters (which wasn't an ideal solution beyond making it easier to copy it)
